lib/std/io_uring.zig

```diff
@@ -0,0 +1,841 @@
+            .sq = sq,
+            .cq = cq,
+            .flags = p.*.flags
+        };
+    }
+
+    pub fn deinit(self: *IO_Uring) void {
+        assert(self.fd >= 0);
+        // The mmaps depend on the fd, so the order of these calls is important:
+        self.cq.deinit();
+        self.sq.deinit();
+        os.close(self.fd);
+        self.fd = -1;
+    }
+
+    /// Returns a vacant SQE, or an error if the submission queue is full.
+    /// We follow the implementation (and atomics) of liburing's `io_uring_get_sqe()` exactly.
+    /// However, instead of a null we return an error to force safe handling.
+    /// Any situation where the submission queue is full tends more towards a control flow error,
+    /// and the null return in liburing is more a C idiom than anything else, for lack of a better
+    /// alternative. In Zig, we have first-class error handling... so let's use it.
+    /// Matches the implementation of io_uring_get_sqe() in liburing.
+    pub fn get_sqe(self: *IO_Uring) !*io_uring_sqe {
+        const head = @atomicLoad(u32, self.sq.head, .Acquire);
+        // Remember that these head and tail offsets wrap around every four billion operations.
+        // We must therefore use wrapping addition and subtraction to avoid a runtime crash.
+        const next = self.sq.sqe_tail +% 1;
+        if (next -% head > self.sq.sqes.len) return error.IO_UringSubmissionQueueFull;
+        var sqe = &self.sq.sqes[self.sq.sqe_tail & self.sq.mask.*];
+        self.sq.sqe_tail = next;
+        return sqe;
+    }
+
+    /// Submits the SQEs acquired via get_sqe() to the kernel. You can call this once after you have
+    /// called get_sqe() multiple times to setup multiple I/O requests.
+    /// Returns the number of SQEs submitted.
+    /// Matches the implementation of io_uring_submit() in liburing.
+    pub fn submit(self: *IO_Uring) !u32 {
+        return self.submit_and_wait(0);
+    }
+
+    /// Like submit(), but allows waiting for events as well.
+    /// Returns the number of SQEs submitted.
+    /// Matches the implementation of io_uring_submit_and_wait() in liburing.
+    pub fn submit_and_wait(self: *IO_Uring, wait_nr: u32) !u32 {
+        var submitted = self.flush_sq();
+        var flags: u32 = 0;
+        if (self.sq_ring_needs_enter(submitted, &flags) or wait_nr > 0) {
+            if (wait_nr > 0 or (self.flags & linux.IORING_SETUP_IOPOLL) > 0) {
+                flags |= linux.IORING_ENTER_GETEVENTS;
+            }
+            return try self.enter(submitted, wait_nr, flags);
+        }
+        return submitted;
+    }
+
+    // Tell the kernel we have submitted SQEs and/or want to wait for CQEs.
+    // Returns the number of SQEs submitted.
+    fn enter(self: *IO_Uring, to_submit: u32, min_complete: u32, flags: u32) !u32 {
+        assert(self.fd >= 0);
+        const res = linux.io_uring_enter(self.fd, to_submit, min_complete, flags, null);
+        try check_errno(res);
+        return @truncate(u32, res);
```

Comment:
I think this should be `@intCast`?

Reply:
I saw you used `@truncate(u32, res)` in your draft PR? The intention is to truncate the `usize` down to a `u32`? I suppose @intCast might give runtime safety in the event that `res` has some unexpected significant bits? Let me know either way...

Reply:
Your call!